app/models/manageiq/providers/amazon/storage_manager/s3/cloud_object_store_object.rb

```diff
@@ -18,6 +18,10 @@ def provider_object(connection = nil)
   end
 
   def raw_delete
-    with_provider_object(&:delete)
+    if key.end_with? "/" # delete object with subobjects (aka. folder)
```

Comment:
Although `/` is probably the most common one, I think this is not completely correct. To my knowledge S3 does not have a concept of directories - the fact that the console supports "creating" them and shows them hierarchically is merely a presentation detail simplifying the workflow.

Experiment I just did. In the S3 console I created a "folder" named `testing` and uploaded a file into it. Listing the buckets using S3 client returns (I truncated irrelevant info)

```ruby
Aws> s3.list_objects({bucket: "miha-oregon"})
[Aws::S3::Client 200 0.943381 0 retries] list_objects(bucket:"miha-oregon",encoding_type:"url")
=> #<struct Aws::S3::Types::ListObjectsOutput
 contents=
  [#<struct Aws::S3::Types::Object
    key="testing/",
    storage_class="STANDARD",
   #<struct Aws::S3::Types::Object
    key="testing/test.rtf",
    storage_class="STANDARD",
```

There are two objects. The fact that the first one is named `testing/` does not make it a parent of `testing/test.rtf`. If I now delete `testing/`

```ruby
Aws> s3.delete_object({bucket: "miha-oregon", key: "testing/"})
[Aws::S3::Client 204 0.879845 0 retries] delete_object(bucket:"miha-oregon",key:"testing/")
=> #<struct Aws::S3::Types::DeleteObjectOutput delete_marker=nil, version_id=nil, request_charged=nil>
```

and list the objects again:

```ruby
Aws> s3.list_objects({bucket: "miha-oregon"})
[Aws::S3::Client 200 1.142258 0 retries] list_objects(bucket:"miha-oregon",encoding_type:"url")
=> #<struct Aws::S3::Types::ListObjectsOutput
 contents=
  [#<struct Aws::S3::Types::Object
    key="testing/test.rtf",
    storage_class="STANDARD",
```

Notice that `testing/` is in fact gone, but `testing/test.rtf` is still there. If you go into the S3 console, you will still see the "folder" and the file within.

This is also the reason why `list_objects` has an additional attribute `delimiter` [doc](http://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Client.html#list_objects-instance_method).

Honestly, I don't know if this is relevant for this function, but it's important to understand this.